File: data-structures/javascript/singly_linked_list.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  // add a new node to the end of the list
  append(value) {
    const newNode = new Node(value);

    if (!this.head) {
      this.head = newNode;
      this.length = 1;
      return;
    }

    const lastNode = this.traverse(this.head);

    lastNode.next = newNode;
    this.length += 1;
  }

  delete(value) {
    if (this.head.value === value) {
      this.head = this.head.next;
    }

    let current = this.head;

    while (current.next && current.next.value !== value) {
      current = current.next;
    }

    if (current.next && current.next.value === value) {
      current.next = current.next.next;
      this.length -= 1;
    }
  }

  traverse(node) {
    if (!node.next) {
      return node;
    }
    return this.traverse(node.next);
  }

  log() {
    let current = this.head;

    const list = [];
    while (current) {
      list.push(current.value);
      current = current.next;
    }

    console.log(list.join("->"));
  }
}

const list = new LinkedList();

list.append(2);
list.append(3);
list.append(4);
list.append(5);
list.append(6);
list.append(7);

list.delete(6);
list.delete(5);

list.log();

console.log(list.length);

File: data-structures/python/singly_linked_list.py

class Node:
    value = None
    next = None
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    head: Node | None = None

    def insert_at_end(self, value):
        newNode = Node(value)
        if self.head is None:
            self.head = newNode
            return
        
        current = self.head
        while current.next:
            current = current.next

        current.next = newNode

    def delete_value(self, value):
        if self.head.value == value:
            self.head = self.head.next

        current = self.head

        while current.next and current.next.value != value:
            current = current.next

        if current.next:
            current.next = current.next.next
            print(f'Deleted value: {value}')

        else:
            print(f'Value: {value} was found in the list')
        

    def print_list(self):
        current = self.head

        while current:
            print(f"{current.value}->")
            current = current.next

ll = LinkedList()

ll.insert_at_end(10)
ll.insert_at_end(20)
ll.insert_at_end(30)
ll.insert_at_end(40)

ll.print_list()
ll.delete_value(20)
ll.delete_value(20)
ll.print_list()